python_modules/dagster/dagster/_components/__init__.py

```diff
@@ -0,0 +1,179 @@
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.definitions_class import Definitions
+
+
+class Component(ABC):
+    name: ClassVar[Optional[str]] = None
+
+    @classmethod
+    def registered_name(cls):
+        return cls.name or to_snake_case(cls.__name__)
+
+    @classmethod
+    def generate_files(cls) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def build_defs(self) -> "Definitions": ...
+
+
+def is_inside_deployment_project(path: str = ".") -> bool:
+    try:
+        _resolve_deployment_root(path)
+        return True
+    except DagsterError:
+        return False
+
+
+def _resolve_deployment_root(path: str) -> str:
+    current_path = os.path.abspath(path)
+    while not _is_deployment_root(current_path):
+        current_path = os.path.dirname(current_path)
+        if current_path == "/":
+            raise DagsterError("Cannot find deployment root")
+    return current_path
+
+
+def is_inside_code_location_project(path: str = ".") -> bool:
+    try:
+        _resolve_code_location_root(path)
+        return True
+    except DagsterError:
+        return False
+
+
+def _resolve_code_location_root(path: str) -> str:
+    current_path = os.path.abspath(path)
+    while not _is_code_location_root(current_path):
+        current_path = os.path.dirname(current_path)
+        if current_path == "/":
+            raise DagsterError("Cannot find code location root")
+    return current_path
+
+
+def _is_deployment_root(path: str) -> bool:
+    return os.path.exists(os.path.join(path, "code_locations"))
+
+
+def _is_code_location_root(path: str) -> bool:
+    return os.path.basename(os.path.dirname(path)) == "code_locations"
+
+
+# Deployment
+_DEPLOYMENT_CODE_LOCATIONS_DIR: Final = "code_locations"
+
+# Code location
+_CODE_LOCATION_CUSTOM_COMPONENTS_DIR: Final = "lib"
+_CODE_LOCATION_COMPONENT_INSTANCES_DIR: Final = "components"
+
+
+class DeploymentProjectContext:
+    def __init__(self, path: Optional[str] = None):
+        path = path or os.getcwd()
+        self._deployment_root = _resolve_deployment_root(path)
+
+    @property
+    def deployment_root(self) -> str:
+        return os.path.join(self._deployment_root, _DEPLOYMENT_CODE_LOCATIONS_DIR)
+
+    @property
+    def code_location_root(self) -> str:
+        return os.path.join(self._deployment_root, _DEPLOYMENT_CODE_LOCATIONS_DIR)
+
+    def has_code_location(self, name: str) -> bool:
+        return os.path.exists(os.path.join(self._deployment_root, "code_locations", name))
+
+
+class CodeLocationProjectContext:
+    def __init__(self, path: Optional[str] = None):
+        path = path or os.getcwd()
+        self._deployment_context = DeploymentProjectContext(path)
+        self._root = _resolve_code_location_root(path)
+        self._name = os.path.basename(self._root)
+        self._component_registry = ComponentRegistry()
+
+        # Make sure we can import from the cwd
+        if sys.path[0] != "":
+            sys.path.insert(0, "")
```

Comment:
hmmm what's up with this. seems fishy

Reply:
Yep see my comment in OP:

>Note also there is some weirdness with sys.path, which by default has "" as first entry (allowing imports from cwd) but which for some reason is missing this entry in the test environment. I've hacked it in with some code that will probably go away once the underyling issue can be identified.

I don't think we should block merging this but I definitely plan to find a better solution upstack.